projects/fmcdaq2/src/app/main.c

```diff
@@ -0,0 +1,818 @@
+		channel_divider = 256;
+		(&p_ad9523_param->channels[ADC_FPGA_CLK])->
+		channel_divider = 4;
+		(&p_ad9523_param->channels[ADC_DEVICE_CLK])->
+		channel_divider = 2;
+		(&p_ad9523_param->channels[ADC_DEVICE_SYSREF])->
+		channel_divider = 256;
+		(&p_ad9523_param->channels[ADC_FPGA_SYSREF])->
+		channel_divider = 256;
+		p_ad9144_param->lane_rate_kbps = 5000000;
+		ad9144_xcvr_param->lane_rate_khz = 5000000;
+#ifndef ALTERA_PLATFORM
+		ad9144_xcvr_param->ref_rate_khz = 250000;
+#else
+		ad9144_xcvr_param->parent_rate_khz = 250000;
+#endif
+		p_ad9680_param->lane_rate_kbps = 5000000;
+		ad9680_xcvr_param->lane_rate_khz = 5000000;
+#ifndef ALTERA_PLATFORM
+		ad9680_xcvr_param->ref_rate_khz = 250000;
+#else
+		ad9680_xcvr_param->parent_rate_khz = 250000;
+#endif
+#ifndef ALTERA_PLATFORM
+		ad9144_xcvr_param->lpm_enable = 1;
+		ad9144_xcvr_param->cpll_enable = 1;
+		ad9144_xcvr_param->out_clk_sel = 4;
+
+		ad9680_xcvr_param->lpm_enable = 1;
+		ad9680_xcvr_param->cpll_enable = 1;
+		ad9680_xcvr_param->out_clk_sel = 4;
+#endif
+		break;
+	case '2':
+		printf ("2 - ADC  500 MSPS; DAC 1000 MSPS\n");
+		p_ad9523_param->pll2_vco_diff_m1 = 3;
+		(&p_ad9523_param->channels[DAC_FPGA_CLK])->
+		channel_divider = 2;
+		(&p_ad9523_param->channels[DAC_DEVICE_CLK])->
+		channel_divider = 1;
+		(&p_ad9523_param->channels[DAC_DEVICE_SYSREF])->
+		channel_divider = 128;
+		(&p_ad9523_param->channels[DAC_FPGA_SYSREF])->
+		channel_divider = 128;
+		(&p_ad9523_param->channels[ADC_FPGA_CLK])->
+		channel_divider = 4;
+		(&p_ad9523_param->channels[ADC_DEVICE_CLK])->
+		channel_divider = 2;
+		(&p_ad9523_param->channels[ADC_DEVICE_SYSREF])->
+		channel_divider = 256;
+		(&p_ad9523_param->channels[ADC_FPGA_SYSREF])->
+		channel_divider = 256;
+		p_ad9144_param->lane_rate_kbps = 10000000;
+		ad9144_xcvr_param->lane_rate_khz = 10000000;
+#ifndef ALTERA_PLATFORM
+		ad9144_xcvr_param->ref_rate_khz = 500000;
+#else
+		ad9144_xcvr_param->parent_rate_khz = 500000;
+#endif
+		p_ad9680_param->lane_rate_kbps = 5000000;
+		ad9680_xcvr_param->lane_rate_khz = 5000000;
+#ifndef ALTERA_PLATFORM
+		ad9680_xcvr_param->ref_rate_khz = 500000;
+#else
+		ad9680_xcvr_param->parent_rate_khz = 500000;
+#endif
+#ifndef ALTERA_PLATFORM
+		ad9144_xcvr_param->lpm_enable = 0;
+		ad9144_xcvr_param->cpll_enable = 0;
+		ad9144_xcvr_param->out_clk_sel = 4;
+
+		ad9680_xcvr_param->lpm_enable = 1;
+		ad9680_xcvr_param->cpll_enable = 1;
+		ad9680_xcvr_param->out_clk_sel = 4;
+#endif
+		break;
+	default:
+		printf ("1 - ADC 1000 MSPS; DAC 1000 MSPS\n");
+#ifndef ALTERA_PLATFORM
+		ad9144_xcvr_param->ref_rate_khz = 500000;
+		ad9680_xcvr_param->ref_rate_khz = 500000;
+#else
+		ad9144_xcvr_param->parent_rate_khz = 500000;
+		ad9680_xcvr_param->parent_rate_khz = 500000;
+#endif
+		break;
+	}
+
+	return(0);
+}
+
+int main(void)
+{
+	int32_t status;
+
+	/* Initialize SPI structures */
+	struct spi_init_param ad9523_spi_param = {
+		.max_speed_hz = 2000000u,
+		.chip_select = 0,
+		.mode = SPI_MODE_0
+	};
```

Comment:
Instead of initializing multiple structures and pass them as a reference, you can declare them inside the main structure.
For example:

```
struct spi_init_param ad9523_spi_param = {
		.max_speed_hz = 2000000u,
		.chip_select = 0,
		.mode = SPI_MODE_0
		.extra = &(struct xil_spi_param) {
				.type = SPI_PL,
				.device_id = SPI_DEVICE_ID
		}
	};
```

This applies for all the structures.